sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs

```diff
@@ -100,28 +135,63 @@ private async Task CommitBlockEvent(StageChunkEventArgs args)
         {
             if (args.Success)
             {
-                Interlocked.Add(ref _bytesTransferred, args.BytesTransferred);
-                // Use progress tracker to get the amount of bytes transferred
-                if (_bytesTransferred == _expectedLength)
-                {
-                    // Add CommitBlockList task to the channel
-                    await _queueCommitBlockTask().ConfigureAwait(false);
-                }
-                else if (_bytesTransferred > _expectedLength)
-                {
-                    await _updateTransferStatus(StorageTransferStatus.CompletedWithSkippedTransfers).ConfigureAwait(false);
-                    await _invokeFailedEventHandler(
-                            new Exception("Unexpected Error: Amount of bytes transferred exceeds expected length.")).ConfigureAwait(false);
-                }
-                _reportProgressInBytes(_bytesTransferred);
+                // Let's add to the channel, and our notifier will handle the chunks.
+                await _stageChunkChannel.Writer.WriteAsync(args, _cancellationToken).ConfigureAwait(false);
             }
             else
             {
-                // Set status to completed
                 await _invokeFailedEventHandler(new Exception("Failure on Stage Block")).ConfigureAwait(false);
             }
         }
 
+        private async Task NotifyOfPendingStageChunkEvents()
+        {
+            try
+            {
+                while (await _stageChunkChannel.Reader.WaitToReadAsync(_cancellationToken).ConfigureAwait(false))
+                {
+                    // Read one event argument at a time.
+                    StageChunkEventArgs args = await _stageChunkChannel.Reader.ReadAsync(_cancellationToken).ConfigureAwait(false);
+                    try
+                    {
+                        await _currentBytesSemaphore.WaitAsync(_cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // We should not continue if waiting on the semaphore has cancelled out.
+                        return;
+                    }
+
+                    Interlocked.Add(ref _bytesTransferred, args.BytesTransferred);
```

Comment:
Take my comment here with a grain of salt since again, not too familiar with .NET concurrency topics but, I'm wondering if we need the Semaphore, Interlocked Add, and the channel all for safely updating `_bytesTransferred`?

Since this is a single reader channel, would there only be one thread executing this code at once? More curious, it's probably fine to leave all of them to be extra safe.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/33049